feat(forms): add constraint validator

jefflembeck opened the pull request:
Form validation needed an overhaul. We don't have proper frontend
validation and we don't necessarily have the time to fold something in.
There should be a singular place to handle when/how errors show and they should
use the constraint validation api in order to do so.

You can opt-in to this functionality by using the `novalidate` attribute
on a form. This seems counterintuitive, but it will force custom
validation handling over what the browser does by itself.

BREAKING CHANGE: This changes the way link updater works, as it
separates the concerns of updating the link underneath and handling the
validation/error showing.

NOTE: There are a lot of updates here. I had to do a lot of weird publishing to
check things out and to make sure it would work.

jefflembeck committed:
feat(forms): add constraint validator

Form validation needed an overhaul. We don't have proper frontend
validation and we don't necessarily have the time to fold something in.
There should be a singular place to handle when/how errors show and they should
use the constraint validation api in order to do so.

You can opt-in to this functionality by using the `novalidate` attribute
on a form. This seems counterintuitive, but it will force custom
validation handling over what the browser does by itself.

BREAKING CHANGE: This changes the way link updater works, as it
separates the concerns of updating the link underneath and handling the
validation/error showing.

aredridel commented on package.json:
```diff
@@ -114,7 +114,7 @@
     "@npmcorp/pui-css-ellipsis": "5.4.0",
     "@npmcorp/pui-css-embeds": "5.4.0",
-    "@npmcorp/pui-css-forms": "5.8.0",
+    "@npmcorp/pui-css-forms": "^6.0.0-alpha.2",
```
this will be updated for final?

jefflembeck replied:
yes absolutely. I'll run it through the normal release process after merging.

aredridel commented on src/pivotal-ui/components/forms/validator.js:
```diff
@@ -0,0 +1,71 @@
+
+var handleBlur = function handleBlur(e) {
+  var input = e.target;
+  if(!input.checkValidity()) {
```
what states does this represent? It looks like a state transition function but it's not naming what it's going from and to

jefflembeck replied:
so, `checkValidity` is fun in that it returns true if you pass any restricting tests that one might have on the element.

Example: Is `required` on the element? If the element is not filled in, `checkValidity()` will return `false`.

You can customize it as well (see `setCustomValidity`)

aredridel replied:
So this is the transition from any state (has error or not) to has error?

jefflembeck replied:
This statement, in particular, is just checking if there is an error on value of the input.

aredridel commented on src/pivotal-ui/components/forms/forms.scss:
```diff
@@ -1470,3 +1470,98 @@ For an organization name input:
+  }
+
+  &.has-error {
+    margin-bottom: 35px;
```
where does this 35px come from?

jefflembeck replied:
Variable, coming right up!

aredridel commented on src/pivotal-ui/components/forms/validator.js:
```diff
@@ -0,0 +1,71 @@
+    if(err) {
+      e.target.setCustomValidity(err.message);
+    } else {
+      e.target.setCustomValidity('');
```
empty string? Come on browser spec makers, you can do better than that!

jefflembeck replied:
riiighhht?

aredridel commented on src/pivotal-ui/components/forms/validator.js:
```diff
@@ -0,0 +1,71 @@
+  if(isLinkUpdater) {
+    var err = isInvalidName(e.target.value);
+    if(err) {
+      e.target.setCustomValidity(err.message);
```
might as well use input instead of `e.target` here, eh? you've got that nice specific alias...

jefflembeck replied:
yup, definite mistake there.

aredridel commented on src/pivotal-ui/components/forms/validator.js:
```diff
@@ -0,0 +1,71 @@
+    }
+  }
+
+  if(!input.checkValidity()) {
```
you've repeated this a couple times: perhaps a function `reflectCurrentValidityState` or similar?

jefflembeck replied:
ooo. dig it.

aredridel commented:
Conceptually this is really good. It should serve us quite well and seems to reflect the trend in how we're designing inputs and doesn't mismatch browser semantics much.

I've some reservations about the `novalidate` opting you in to validating behavior. Perhaps that combined with something that specifically triggers our validation system? Or have this add novalidate itself if it's turned on for an element?

jefflembeck commented:
@aredridel yeah, i've been thinking about that since I said it out loud. I think you're totally right. Opt-in with `data-validate` or something to that effect and that'll take care of the rest.

jefflembeck committed:
fix(forms): refactor and opt-in on validation

Use data-validate to turn on our in-house validation. Yay.

Also, break out a function for reuse and make sure to stay away from
magic numbers

aredridel commented on src/pivotal-ui/components/forms/forms.scss:
````diff
@@ -1549,6 +1551,8 @@ the way we see fit.
 ```
 */
 
+$form-input-error-height: 35px;
````
So where does _this_ 35px come in? Why is it 35?

jefflembeck replied:
Honestly? Working with jerry on making errors show nicely. Kept playing with it until we hit 35px

aredridel replied:
Fair enough! I'd figured it was probably a function of the font-size and line height but wasn't sure.

aredridel commented:
🐑 it!
